src/shared/components/oncoprint/OncoprintUtils.ts

```diff
@@ -531,6 +538,208 @@ export function getCategoricalTrackRuleSetParams(
     };
 }
 
+/**
+ * Processes molecular data for heatmap tracks based on profile type.
+ * For MUTATION_EXTENDED profiles, calculates variant allele frequency (VAF) from mutations.
+ * For other profiles, retrieves molecular data values directly from cache.
+ */
+function createHeatmapTracksData(
+    query: any,
+    profileType: string,
+    oncoprint: ResultsViewOncoprint,
+    useOqlFiltering: boolean = false
+): Array<{
+    uniqueSampleKey: string;
+    uniquePatientKey: string;
+    value: number | null;
+    thresholdType?: '>' | '<';
+}> {
+    if (profileType === AlterationTypeConstants.MUTATION_EXTENDED) {
+        const samples = oncoprint.props.store.filteredSamples.result!;
+        const coverageInfo = oncoprint.props.store.coverageInformation.result!;
+        const queriedGenes = oncoprint.props.store.genes.result || [];
+        const isGeneInQuery = queriedGenes.some(
+            g =>
+                g.hugoGeneSymbol.toUpperCase() ===
+                query.hugoGeneSymbol.toUpperCase()
+        );
+
+        let mutations: any[] | undefined;
+
+        if (
+            useOqlFiltering &&
+            oncoprint.props.store.oqlFilteredAlterations.isComplete
+        ) {
+            const oqlFiltered =
+                oncoprint.props.store.oqlFilteredAlterations.result;
+            if (oqlFiltered) {
+                const geneSymbol = query.hugoGeneSymbol.toUpperCase();
+                mutations = oqlFiltered.filter(
+                    (alt: any) =>
+                        alt.hugoGeneSymbol &&
+                        alt.hugoGeneSymbol.toUpperCase() === geneSymbol &&
+                        alt.molecularProfileAlterationType ===
+                            'MUTATION_EXTENDED'
+                );
+            }
+        } else {
+            const mutationPromise = oncoprint.props.store.annotatedMutationCache.get(
+                {
+                    entrezGeneId: query.entrezGeneId,
+                }
+            );
+            mutations = mutationPromise.result;
+            if (mutations) {
+                mutations = mutations.filter(
+                    m => m.molecularProfileId === query.molecularProfileId
+                );
+            }
+        }
+
+        if (mutations && mutations.length > 0) {
+            const sampleMutationMap = new Map<string, number | null>();
```

Comment:
a value of `null` is a bit confusing. we should probably use some other constant value here.